src/main/java/gregtech/common/metatileentities/storage/CraftingRecipeLogic.java

```diff
@@ -0,0 +1,198 @@
+                craftingGridChanged = true;
+            }
+        }
+        return craftingGridChanged;
+    }
+
+    public void performRecipe(EntityPlayer player) {
+        if (!isRecipeValid()) {
+            return;
+        }
+        if (!cachedRecipeData.consumeRecipeItems()) {
+            return;
+        }
+        ForgeHooks.setCraftingPlayer(player);
+        NonNullList<ItemStack> remainingItems = cachedRecipe.getRemainingItems(inventoryCrafting); // todo right here is where tools get damaged (in UI)
+        ForgeHooks.setCraftingPlayer(null);
+        for (int i = 0; i < remainingItems.size(); i++) {
+            ItemStack itemStack = remainingItems.get(i);
+            if (itemStack.isEmpty()) {
+                continue;
+            }
+            ItemStackKey stackKey = KeySharedStack.getRegisteredStack(itemStack);
+            inventoryCrafting.setInventorySlotContents(i, itemStack);
+            int remainingAmount = itemStack.getCount() - itemSources.insertItem(stackKey, itemStack.getCount(), false, IItemList.InsertMode.HIGHEST_PRIORITY);
+            if (remainingAmount > 0) {
+                itemStack.setCount(remainingAmount);
+                player.addItemStackToInventory(itemStack);
+                if (itemStack.getCount() > 0) {
+                    player.dropItem(itemStack, false, false);
+                }
+            }
+        }
+    }
+
+    public void handleItemCraft(ItemStack itemStack, EntityPlayer player, boolean simulate) {
+        itemStack.onCrafting(world, player, 1);
+        itemStack.getItem().onCreated(itemStack, world, player);
+        if (!simulate) {
+            //if we're not simulated, fire the event, unlock recipe and add crafted items, and play sounds
+            FMLCommonHandler.instance().firePlayerCraftingEvent(player, itemStack, inventoryCrafting);
+
+            if (cachedRecipe != null && !cachedRecipe.isDynamic()) {
+                player.unlockRecipes(Lists.newArrayList(cachedRecipe));
+            }
+            if (cachedRecipe != null) {
+                ItemStack resultStack = cachedRecipe.getCraftingResult(inventoryCrafting);
+                this.itemsCrafted += resultStack.getCount();
+                recipeMemory.notifyRecipePerformed(craftingGrid, resultStack);
+            }
+        }
+    }
+
+    public void refreshOutputSlot() {
+        ItemStack itemStack = ItemStack.EMPTY;
+        if (cachedRecipe != null) {
+            itemStack = cachedRecipe.getCraftingResult(inventoryCrafting);
+        }
+        this.craftingResultInventory.setInventorySlotContents(0, itemStack);
+    }
+
+    public boolean isRecipeValid() {
+        return cachedRecipeData.attemptMatchRecipe() == 511;
+    }
+
+    private void updateCurrentRecipe() {
+        if (!cachedRecipeData.matches(inventoryCrafting, world)) {
+            IRecipe newRecipe = CraftingManager.findMatchingRecipe(inventoryCrafting, world);
```

Comment:
Should we check the memorized recipes first before iterating through the registry?